main.go

```diff
@@ -1,115 +1,144 @@
+	"github.com/cloudfoundry/noaa/consumer"
 	"github.com/quipo/statsd"
 	"gopkg.in/alecthomas/kingpin.v2"
 )
 
 var (
-	dopplerEndpoint   = kingpin.Flag("doppler-endpoint", "Doppler endpoint").Default("wss://doppler.10.244.0.34.xip.io:443").OverrideDefaultFromEnvar("DOPPLER_ENDPOINT").String()
-	uaaEndpoint       = kingpin.Flag("uaa-endpoint", "UAA endpoint").Default("https://uaa.10.244.0.34.xip.io").OverrideDefaultFromEnvar("UAA_ENDPOINT").String()
-	subscriptionId    = kingpin.Flag("subscription-id", "Id for the subscription.").Default("firehose").OverrideDefaultFromEnvar("SUBSCRIPTION_ID").String()
+	apiEndpoint       = kingpin.Flag("api-endpoint", "API endpoint").Default("https://api.10.244.0.34.xip.io").OverrideDefaultFromEnvar("API_ENDPOINT").String()
 	statsdEndpoint    = kingpin.Flag("statsd-endpoint", "Statsd endpoint").Default("10.244.11.2:8125").OverrideDefaultFromEnvar("STATSD_ENDPOINT").String()
 	statsdPrefix      = kingpin.Flag("statsd-prefix", "Statsd prefix").Default("mycf.").OverrideDefaultFromEnvar("STATSD_PREFIX").String()
-	prefixJob         = kingpin.Flag("prefix-job", "Prefix metric names with job.index").Default("false").OverrideDefaultFromEnvar("PREFIX_JOB").Bool()
-	username          = kingpin.Flag("username", "Firehose username.").Default("admin").OverrideDefaultFromEnvar("FIREHOSE_USERNAME").String()
-	password          = kingpin.Flag("password", "Firehose password.").Default("admin").OverrideDefaultFromEnvar("FIREHOSE_PASSWORD").String()
+	username          = kingpin.Flag("username", "UAA username.").Default("").OverrideDefaultFromEnvar("USERNAME").String()
+	password          = kingpin.Flag("password", "UAA password.").Default("").OverrideDefaultFromEnvar("PASSWORD").String()
 	skipSSLValidation = kingpin.Flag("skip-ssl-validation", "Please don't").Default("false").OverrideDefaultFromEnvar("SKIP_SSL_VALIDATION").Bool()
 	debug             = kingpin.Flag("debug", "Enable debug mode. This disables forwarding to statsd and prints to stdout").Default("false").OverrideDefaultFromEnvar("DEBUG").Bool()
+	updateFrequency   = kingpin.Flag("update-frequency", "The time in seconds, that takes between each apps update call.").Default("300").OverrideDefaultFromEnvar("UPDATE_FREQUENCY").Int64()
+	metricTemplate    = kingpin.Flag("metric-template", "The template that will form a new metric namespace.").Default("{{.Space}}.{{.App}}.{{.Metric}}").OverrideDefaultFromEnvar("METRIC_TEMPLATE").String()
 )
 
 func main() {
 	kingpin.Parse()
 
-	tokenFetcher := &token.UAATokenFetcher{
-		UaaUrl:                *uaaEndpoint,
-		Username:              *username,
-		Password:              *password,
-		InsecureSSLSkipVerify: *skipSSLValidation,
+	c := &cfclient.Config{
+		ApiAddress:        *apiEndpoint,
+		SkipSslValidation: *skipSSLValidation,
+		Username:          *username,
+		Password:          *password,
 	}
 
-	authToken, err := tokenFetcher.FetchAuthToken()
+	client, err := cfclient.NewClient(c)
```

Comment:
I managed to reproduce the problem. To make it fail, you need to change the UAA `cf` client settings with this command:

```
uaac target https://api.${DEPLOY_ENV}.dev.cloudpipeline.digital --skip-ssl-validation
uaac token client get admin -s $(aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cf-secrets.yml -  | awk '/uaa_admin_client_secret/ {print $2}')

# and change token settiings
uaac client update cf --access_token_validity 3 --refresh_token_validity 6
```

Then it would fail:
```
$ go run main.go  --api-endpoint https://api.${DEPLOY_ENV}.dev.cloudpipeline.digital --skip-ssl-validation --statsd-endpoint localhost:8125 --username admin --password $(aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cf-secrets.yml -  | awk '/uaa_admin_password/ {print $2}')  --metric-template="{{.Space}}.{{.App}}.{{.Instance}}.{{.Metric}}" --update-frequency=2
Error getting bearer token: oauth2: cannot fetch token: 401 Unauthorized
Response: {"error":"invalid_token","error_description":"The token expired, was revoked, or the token ID is incorrect: f072e5b066d84d3b82ecdd920d663493-r"}
exit status 255
```

We tried a patch that re-authenticates if the token is invalid

```diff
diff --git a/main.go b/main.go
index 6c84575..f104bd9 100644
--- a/main.go
+++ b/main.go
@@ -4,6 +4,7 @@ import (
        "crypto/tls"
        "fmt"
        "os"
+       "strings"
        "time"
 
        "github.com/alphagov/paas-cf-apps-statsd/metrics"
@@ -36,12 +37,6 @@ func main() {
                Password:          *password,
        }
 
-       client, err := cfclient.NewClient(c)
-       if err != nil {
-               fmt.Println(err)
-               os.Exit(-1)
-       }
-
        containerMetricProcessor := processors.NewContainerMetricProcessor()
 
        sender := statsd.NewStatsdClient(*statsdEndpoint, *statsdPrefix)
@@ -61,15 +56,28 @@ func main() {
 
        go func() {
                apps := make(map[string]*consumer.Consumer)
-
                for {
-                       err := updateApps(client, apps, msgChan, errorChan)
+
+                       fmt.Println("Authentication")
+                       client, err := cfclient.NewClient(c)
                        if err != nil {
                                fmt.Println(err)
                                os.Exit(-1)
                        }
 
-                       time.Sleep(time.Duration(*updateFrequency) * time.Second)
+                       for {
+                               err := updateApps(client, apps, msgChan, errorChan)
+                               if err != nil {
+                                       if strings.Contains(err.Error(), `"error":"invalid_token"`) {
+                                               fmt.Println("Refresh Token has expired")
+                                               break
+                                       }
+                                       fmt.Println(err)
+                                       os.Exit(-1)
+                               }
+
+                               time.Sleep(time.Duration(*updateFrequency) * time.Second)
+                       }
                }
        }()
 
@@ -140,5 +148,17 @@ func updateApps(client *cfclient.Client, watchedApps map[string]*consumer.Consum
                }
        }
 
+       authToken, err = client.GetToken()
+       if err != nil {
+               return err
+       }
+
+       fmt.Printf("Bearer Token: %v\n", authToken)
+       authRefreshToken, err := client.GetRefreshToken()
+       if err != nil {
+               return err
+       }
+       fmt.Printf("Refresh Token: %v\n", authRefreshToken)
+
        return nil
 }
``` 
```